state/optimistic_syncing/pipeline.go

```
@@ -0,0 +1,474 @@
+
+	mu                 sync.RWMutex
+	state              State
+	isSealed           bool
+	descendsFromSealed bool
+	parentState        State
+	executionResultID  string
+	blockID            string
+
+	core          Core
+	stateNotifier engine.Notifier
+}
+
+// NewPipeline creates a new processing pipeline.
+// The pipeline is always initialized in the Ready state and will begin processing
+// when Run is called.
+func NewPipeline(config Config) *Pipeline {
+	p := &Pipeline{
+		logger:              config.Logger.With().Str("component", "pipeline").Str("executionResultID", config.ExecutionResultID).Logger(),
+		stateUpdateChan:     make(chan StateUpdate, 10),
+		childrenUpdateChans: config.ChildrenStateUpdateChans,
+		state:               StateReady,
+		isSealed:            config.IsSealed,
+		descendsFromSealed:  true,
+		stateNotifier:       engine.NewNotifier(),
+		core:                config.Core,
+		executionResultID:   config.ExecutionResultID,
+		blockID:             config.BlockID,
+	}
+
+	return p
+}
+
+// Run starts the pipeline processing and blocks until completion or context cancellation.
+//
+// This function handles the progression through the pipeline states, executing the appropriate
+// processing functions at each step. It also listens for state updates from the parent pipeline
+// and cancels processing if the pipeline no longer descends from the latest persisted result.
+//
+// When the pipeline reaches a terminal state (StateComplete or StateCanceled), the function returns.
+// The function will also return if the provided context is canceled.
+//
+// Returns an error if any processing step fails or if the context is canceled.
+func (p *Pipeline) Run(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	go p.listenForStateUpdates(ctx, cancel)
+
+	notifierChan := p.stateNotifier.Channel()
+
+	// Trigger initial check
+	p.stateNotifier.Notify()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
```

I don't think we need to return an error here. The caller will know that the pipeline completed successfully from it's state updates.